### ISSUE TITLE
Original Date/Time is changed when Copy / View

### ISSUE THREAD

**Issue by TTT7275** [macos-fuse-t/fuse-t#17]
Thanks for create this great module,

I have a problem with Original Date/Time is changed when Copy / Viewing. I created an issue in Cryptomator also -> 
https://github.com/cryptomator/cryptomator/issues/2700   

Here is what I got : 

**When Copy**
- **Cryptomator** - Some file do change the date/time some file do not. I still can't find the pattern since both files have the same pattern except it has different date time value in the file.
- **rclone** - All the files keep the original date/time

**When Preview in Finder according to julesallen**
Both **rclone** and **Cryptomator** do change date/time to current date time.

**Comment by macos-fuse-t** [comment]
Unfortunately this is a bug of macOS that causes Preview to change file modification time (mtime) on an NFS volume

**Comment by TTT7275** [comment]
Then, we can only wait and hope they will fix it.

Thank you for your quick reply.

**Comment by julesallen** [comment]
Hey Alex @macos-fuse-t , thank you for all the work you've put in on getting the project this far. kexts are a pia!

> Unfortunately this is a bug of macOS that causes Preview to change file modification time (mtime) on an NFS volume

I'm not a filesystem developer and don't know where to dig to help get this fixed. You say it's a bug in macOS, is the core OS using an old or buggy version of NFS? Has this been reported to Apple and is the core of the NFS open source? 

Looks like we're not alone here! https://apple.stackexchange.com/questions/452957/mac-quick-preview-change-file-datetime-in-nfs-shares

Happy to do some grunt work to help get this solved if you need the help.

**Comment by macos-fuse-t** [comment]
I will report this bug although I'm pretty sure it was already reported in the past
Perhaps this is the culprit: nfs_vattr_set_bitmap() func:
https://opensource.apple.com/source/xnu/xnu-7195.60.75/bsd/nfs/nfs4_subs.c.auto.html

**Comment by macos-fuse-t** [comment]
A workaround is available in 1.0.22

**Comment by kapitainsky** [comment]
It is great news that `nomtime` and `noatime` has been added!

For completeness could you also add `nodiratime` at some stage? It would solve the same problem for directories.

**Comment by macos-fuse-t** [comment]
`nomtime` and `noatime` apply both to files and folders

**Comment by kapitainsky** [comment]
I have just tested and it does not seem to be the case - here is some folder and file on FUSE-T share:

<img width="425" alt="image" src="https://github.com/macos-fuse-t/fuse-t/assets/15635386/c4019705-b579-4052-a04d-c33534f5f77a">

I copy it to other location on share drive:

<img width="608" alt="image" src="https://github.com/macos-fuse-t/fuse-t/assets/15635386/7b1ff5ab-f8f1-4001-a2e4-695cf5d62137">

I copy it to local disk:

<img width="422" alt="image" src="https://github.com/macos-fuse-t/fuse-t/assets/15635386/416c2926-2390-4a99-9066-0a8323e16bb2">

**Comment by macos-fuse-t** [comment]
How did you test?

**Comment by kapitainsky** [comment]
Also some reference to other project findings when dealing with NFS on macOS

https://github.com/me-io/docker-machine-mac-nfs-mount/blob/master/README.md

**Comment by kapitainsky** [comment]
> How did you test?

fuse-t with `-o noatime` and then I simply use finder copy/paste

**Comment by macos-fuse-t** [comment]
You should use with -o nomtime. noatime is for access time

**Comment by macos-fuse-t** [comment]
Try with a libfuse sample fusexmp_fh which would mirror your hard disk.

**Comment by kapitainsky** [comment]
> You should use with -o nomtime. noatime is for access time

Yes - but this is exactly what I want - to have mtime and prevent its changes by software like Finder trying to change atime.

**Comment by macos-fuse-t** [comment]
Correct, that's why you should mount with "-o nomtime". It won't change modify time when hitting Preview. Access time would still get updated.

**Comment by kapitainsky** [comment]
Yes it wont change - ever:) I want to know when I edited my file... But yes I do understand that people might have different requirement so great we have both options available.

**Comment by macos-fuse-t** [comment]
Use experimental "-o backend=smb" which doesn't have this issue. I don't know what Mountain Duck is or how it works.
See here https://github.com/macos-fuse-t/fuse-t/releases/tag/1.0.35

**Comment by alexanderharm** [comment]
does the smb-option really work? i still receive current datetimes wenn `cp/rsync -a` a file into the mountpoint...

**Comment by jknockaert** [comment]
I am seeing this issue (preview causing the modification timestamp to be updated) on Tahoe (26.5) when using sshfs in combination with the fskit backend. So it is not limited to NFS.

**Comment by macos-fuse-t** [comment]
It's actually a bug in the sshfs implementation. I (codex) updated the code and you can re-install it from brew
```brew install macos-fuse-t/homebrew-cask/sshfs-fuse-t```

**Comment by jknockaert** [comment]
@macos-fuse-t I updated sshfs and the bug is gone -- thanks.
I however hit another bug, where after browsing a couple of subfolders and previewing some files all of a sudden folders are shown as empty. A bit later sshfs is terminated (`by signal SIGABRT (Abort)`):
`Assertion failed: (offset == 0), function cache_readdir, file cache.c, line 504`
I do not user any options other than `backend` and `volname`.

**Comment by macos-fuse-t** [comment]
I wanted to speed up sshfs and messed up. I removed all the optimizations for now, so it should be safe and stable

**Comment by jknockaert** [comment]
@macos-fuse-t I checked the new version (upgrading through homebrew was unsuccessful so I re-installed `sshfs-fuse-t`). Still I got the same arror (now at line 389 instead of 504), even when it took a while longer until the `SIGABRT` without the optimisations.

**Comment by macos-fuse-t** [comment]
This is strange, the current version is almost identical to the latest upstream commit
https://github.com/libfuse/sshfs/commit/9e35c39ba83f54a49a9df4bf0a629f26c60cc38c
Can you build it from the sources https://github.com/macos-fuse-t/sshfs/tree/libfuse3 and check again? Did it work before the mtime changes?

**Comment by jknockaert** [comment]
@macos-fuse-t I haven't been running sshfs for quite a while so I can't confirm how it worked before. I will see if I manage to build it.

**Comment by jknockaert** [comment]
Using a locally built version I get the same result.
The bug may have been in sshfs already a long time. When browsing through (large) subfolders after a while I only get to see empty folders. And then after waiting some more sshfs is terminated by a SIGABRT.
